File: backend/src/commands/stats.rs
use tauri::State;

use crate::models::{StatsSummary, Track, YearlyWrap};
use crate::state::AppState;

#[tauri::command]
pub fn get_stats(state: State<'_, AppState>) -> Result<StatsSummary, String> {
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.get_stats_summary()
}

#[tauri::command]
pub fn get_yearly_wrap(state: State<'_, AppState>, year: i32) -> Result<YearlyWrap, String> {
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.get_yearly_wrap(year)
}

#[tauri::command]
pub fn get_recent_tracks(state: State<'_, AppState>, limit: Option<usize>) -> Result<Vec<Track>, String> {
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.get_recent_tracks(limit.unwrap_or(50))
}

#[tauri::command]
pub fn clear_play_history(state: State<'_, AppState>) -> Result<(), String> {
    let mut db = state.db.lock().map_err(|e| e.to_string())?;
    db.clear_play_history()
}

File: frontend/api/stats.ts
import { invoke } from "@tauri-apps/api/core";

import type { StatsSummary, Track, YearlyWrap } from "../types";
import { requireTauri } from "./client";

export async function getStats(): Promise<StatsSummary> {
  requireTauri();
  return invoke<StatsSummary>("get_stats");
}

export async function getStatsYears(): Promise<number[]> {
  requireTauri();
  return invoke<number[]>("get_stats_years");
}

export async function getYearlyWrap(year: number): Promise<YearlyWrap> {
  requireTauri();
  return invoke<YearlyWrap>("get_yearly_wrap", { year });
}

export async function getRecentTracks(limit = 50): Promise<Track[]> {
  requireTauri();
  return invoke<Track[]>("get_recent_tracks", { limit });
}

export async function clearPlayHistory(): Promise<void> {
  requireTauri();
  return invoke<void>("clear_play_history");
}

export function formatListenDuration(secs: number): string {
  if (!Number.isFinite(secs) || secs < 60) {
    return `${Math.round(secs || 0)} sec`;
  }
  const h = Math.floor(secs / 3600);
  const m = Math.floor((secs % 3600) / 60);
  if (h > 0) return `${h} hr ${m} min`;
  return `${m} min`;
}
